app/controllers/provider_interface/offer_changes_controller.rb

```diff
@@ -0,0 +1,159 @@
+        render_courses
+      else
+        render_course_options
+      end
+    end
+
+    def update
+      @change_offer_form.step = :update
+      if @change_offer_form.valid?
+        ChangeOffer.new(
+          actor: current_provider_user,
+          application_choice: @application_choice,
+          course_option_id: @change_offer_form.course_option_id,
+        ).save!
+        redirect_to provider_interface_application_choice_path(@application_choice.id)
+      else
+        raise 'cannot update offer'
+      end
+    end
+
+  private
+
+    def render_providers
+      set_alternative_providers
+      render :edit_provider
+    end
+
+    def render_courses
+      set_alternative_courses
+      render :edit_course
+    end
+
+    def render_course_options
+      set_alternative_course_options
+      render :edit_course_option
+    end
+
+    def requires_provider_change_response_feature_flag
+      raise unless FeatureFlag.active?('provider_change_response')
+    end
+
+    def set_application_choice
+      @application_choice = GetApplicationChoicesForProviders.call(providers: current_provider_user.providers)
+        .find(params[:application_choice_id])
+    end
+
+    def set_change_offer_form
+      @change_offer_form = ProviderInterface::ChangeOfferForm.new application_choice: @application_choice,
+                                                                  provider_id: (provider.id if allowed_provider?),
+                                                                  course_id: course&.id,
+                                                                  course_option_id: course_option&.id
+    end
+
+    def allowed_provider?
+      current_provider_user.providers.include?(provider)
+    end
+
+    def provider
+      requested_provider || current_provider
+    end
+
+    def current_provider
+      @application_choice.offered_course.provider
+    end
+
+    def requested_provider
+      provider_id = change_offer_params[:provider_id]
+      Provider.find(provider_id) if provider_id.present?
+    end
+
+    def course
+      requested_course || current_course
+    end
+
+    def current_course
+      @application_choice.offered_course
+    end
+
+    def requested_course
+      course_id = change_offer_params[:course_id]
+      Course.find(course_id) if course_id.present?
+    end
+
+    def course_option
+      requested_course_option || current_course_option
+    end
+
+    def current_course_option
+      @application_choice.offered_option
+    end
+
+    def requested_course_option
+      course_option_id = change_offer_params[:course_option_id]
+      CourseOption.find(course_option_id) if course_option_id.present?
+    end
+
+    def set_alternative_providers
+      @alternative_providers = current_provider_user.providers.order(:name)
+    end
+
+    def set_alternative_courses
+      @alternative_courses = Course.where(open_on_apply: true, provider: provider).order(:name)
```

Comment:
Should we exclude the currently selected course? What happens if I just select the same course again?

Reply:
It could have a different study_mode, maybe?

Reply:
If we were worried about providers ending up with an unsatisfying no-op, we could check at the end whether they were attempting to change the offered course to the existing offered course

Reply:
It's a good point, have just pushed a commit introducing this check in the service.

Reply:
Good catch, I've also added a `study_mode` check within `#set_alternative_courses`. I think the code did the right thing without this already, as during the last step it was filtering out study modes different to the current one. But it could have led to presenting courses that, if selected, would have produced empty option/locations lists to choose from in the next step.

Reply:
If we need a separate feature flag for withdrawing applications, I'll refactor then.